### ISSUE TITLE
Sometimes Need to Restart Controller/Destination?

### ISSUE THREAD

**Issue by steve-gray** [linkerd/linkerd2#4855]
### What is the issue?
When a cluster is having issues, sometimes we've observed that linkerd seems to either stop working or stop tracking cluster updates. This means that when looking at pods that are having TCP connection startup issues, we see the following issues. linkerd does not seem to recover without a manual intervention.

Visually on a timeline, we see connection error 111's from the proxy, which clear when we restart linkerd's control services.
![image](https://user-images.githubusercontent.com/14563775/89754024-a3641080-db1d-11ea-8c4a-7c1aa5b15c32.png)

### How can it be reproduced?
Not sure.

### Logs, error output, etc
```
[ 84.119814306s] WARN inbound:accept{peer.addr=10.202.178.154:51116}:source{target.addr=10.202.180.238:8080}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)\n
```

When the issue occurs, the issue is that proxies all seem to get consistently stuck with an old set of addresses to work with. This becomes progressively more disruptive to our system as services restart. Checking pod histories, we can see that the IP/port combinations were valid in past, but were no longer valid at the time of the logs.

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-addons
--------------
√ 'linkerd-config-addons' config map exists

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

### Environment

- Kubernetes Version: 1.16
- Cluster Environment: Amazon EKS
- Host OS: EKS AMI 
- Linkerd version: 

### Possible solution
We can mitigate/clear the issue near-instantly by running:

```kubectl rollout restart -n linkerd deployment/linkerd-controller deployment/linkerd-destination```

### Additional context
It seems if ever the cluster is under distress for any reason, potentially the watch for pod status updates does not get re-established, so that the system stops tracking endpoints, causing the proxies to use their list of last known good endpoints.

**Comment by ihcsim** [comment]
@adleong Does this sound like the staleness issue described in https://github.com/linkerd/linkerd2/issues/3854?

**Comment by zaharidichev** [comment]
> When a cluster is having issues

@steve-gray What issues in particular. Could it be the case that the dst service is restarted or the node where it is dies and the svc is brought up on another node. If that is the case, then I think this is a known issue that we are currently working on

**Comment by steve-gray** [comment]
Basically if the cluster is in CPU distress and starts rescheduling stuff. It _**may**_ be that the linkerd-destination pod goes cactus. Guessing in that scenario the internal watches stop tracking pod changes? Is there a specific log entry I can use to to validate this next time it happens.

Our UAT environment is a very bruised, beaten creature, and we can probably reliably get it happening in the next few days - but would be good to know what logs or indicators to look for? Also once I've collected those, do I just restart the destination pod - or the controller too (or both?)

**Comment by zaharidichev** [comment]
That actually sounds like https://github.com/linkerd/linkerd2/issues/4674

**Comment by adleong** [comment]
This could be #3854 or #4674 or something else.  The easiest thing to check would probably be to test @zaharidichev's fix as described in #4674 and see if that resolves the issue.
